lib/ansible/galaxy/collection.py

```diff
@@ -1046,10 +1048,13 @@ def _extract_tar_file(tar, filename, b_dest, b_temp_path, expected_hash=None):
         if not os.path.exists(b_parent_dir):
             # Seems like Galaxy does not validate if all file entries have a corresponding dir ftype entry. This check
             # makes sure we create the parent directory even if it wasn't set in the metadata.
-            os.makedirs(b_parent_dir)
+            os.makedirs(b_parent_dir, mode=0o0755)
 
         shutil.move(to_bytes(tmpfile_obj.name, errors='surrogate_or_strict'), b_dest_filepath)
 
+        tar_member = tar.getmember(to_native(filename, errors='surrogate_or_strict'))
+        os.chmod(b_dest_filepath, tar_member.mode)
```

Comment:
Do we want to alter the mode here, masking or possibly adding bits? A few things that come to mind:

* We may not want to allow things like `stat.S_ISUID` or `stat.S_ISGID` (in case it's run as root).
* Would lacking `stat.S_IWUSR` for example prevent the files from being removed/updated during uninstall/upgrade?

Reply:
Fair enough, I've set it to do `0o0644` as a default and changed to `0o0755` if the existing tar file has `S_IXUSR`. This extra permissions like `S_ISUID` or just weird permissions in the tar are ignored in case they are trying to do something nefarious.